File: src/mtglab/api/adminstats.py
"""The admin dashboard's numbers: four read-only views of this instance.

Everything here lives under `auth.ADMIN_PREFIX`, exactly as `api/admin.py`
does and for the same reason: the middleware refuses the whole prefix to a
non-admin **before routing** (ADR 17), every handler also depends on
`deps.Admin`, and every route is classified in `tests/test_isolation.py` —
the suite fails until it is.

All four views are facts about this box, read from the box: the process,
the filesystem the volume is mounted on, `app.db`, and the in-memory job
registry. No external API and no new secret.

**No invoice — an estimate, and the difference is the whole design.** Decided
with Aaron 2026-08-18: his Console account is individual, and Anthropic's Usage
& Cost Admin API exists only for organizations, so *the real bill is not
reachable from here and never will be by this route*. That has not changed.
What changed later the same day is that the page stopped refusing to say
anything about money: `claude/prices.py` multiplies the ledger's tokens by list
rates a person read on a dated day, which is an estimate and is labelled one
everywhere it appears. It remains a **floor** — cache writes bill at 1.25x
input and are captured nowhere — and a conversation whose model carries no rate
is counted rather than priced at zero, so the figure is never quietly wrong
downward. Adding money to the account stays a human act; the page links out
rather than pretending.

One absence survives:

- **No machine-level metrics from outside.** Fly's managed Prometheus
  (edge traffic, instance memory as the platform sees it) came later, behind a
  read-only token in `api/flymetrics.py`; *this* module still reports only what
  the process can see without asking anybody.
"""

from __future__ import annotations

import logging
import os
import shutil
import sqlite3
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import Any

from fastapi import FastAPI

from mtglab import config
from mtglab.api import jobs
from mtglab.api.deps import Admin
from mtglab.auth import db, tokens, users
from mtglab.claude import prices, tiers

_LOG = logging.getLogger("mtglab.api.adminstats")


# ------------------------------------------------------------------ helpers

def _size_of(path: Path) -> int | None:
    """Bytes on disk, or None for "nothing there" — a fresh instance has no
    pool and an unseeded cache, and the dashboard should say absent rather
    than zero, which reads as "present and empty"."""
    try:
        if path.is_file():
            return path.stat().st_size
        if path.is_dir():
            return sum(f.stat().st_size for f in path.rglob("*") if f.is_file())
    except OSError as exc:
        _LOG.warning("could not size %s (%s)", path, exc)
    return None


def _cache_breakdown(cache: Path, total: int | None) -> dict[str, int | None]:
    """The cache, tenant by tenant, with whatever is left over named as such.

    Three shelves live under `data/cache` on a deployed volume: the mana
    symbols (ADR 33), the card-motion derivatives (ADR 32) and the reading
    engine (`ocr.py`). `other_bytes` is what the named three do not account
    for, and it is the part that matters: this panel shipped naming two of
    the three while the reading engine was **38% of the cache**, so the tile
    read as a breakdown while quietly under-reporting its newest tenant. A
    remainder cannot hide a fourth shelf the way a fixed list can -- the next
    one somebody adds shows up as unexplained bytes rather than as nothing.
    """
    named = {"symbols_bytes": _size_of(cache / "symbols"),
             "cardmotion_bytes": _size_of(cache / "cardmotion"),
             "ocr_bytes": _size_of(cache / "ocr")}
    other = (None if total is None
             else max(0, total - sum(v for v in named.values() if v)))
    return {**named, "other_bytes": other}


def _count_dirs(path: Path) -> int:
    try:
        return sum(1 for p in path.iterdir() if p.is_dir()) if path.is_dir() else 0
    except OSError:
        return 0


def _rss() -> dict[str, Any]:
    """This process's resident memory, as well as the platform tells it.

    Deployed — the audience that matters — this is Linux, and
    `/proc/self/status` reports the *current* RSS. The dev Mac has no /proc,
    so it falls back to `getrusage`, whose `ru_maxrss` is the *peak* (and in
    bytes there, not kilobytes — the units differ per platform and both
    branches say which). `kind` rides along so the page can label the number
    honestly instead of showing a peak as a level.
    """
    proc = Path("/proc/self/status")
    if proc.exists():
        try:
            for line in proc.read_text().splitlines():
                if line.startswith("VmRSS:"):
                    return {"bytes": int(line.split()[1]) * 1024,
                            "kind": "current"}
        except (OSError, ValueError, IndexError) as exc:
            _LOG.warning("could not read /proc/self/status (%s)", exc)
    import resource
    peak = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    # Linux reports kilobytes, macOS bytes. If /proc was unreadable we are
    # probably not on Linux, but check the platform rather than assume.
    import sys
    factor = 1 if sys.platform == "darwin" else 1024
    return {"bytes": peak * factor, "kind": "peak"}


def _machine_memory() -> dict[str, Any]:
    """Total physical memory, and — where the kernel says — what is free.

    `MemAvailable` is Linux's own answer to "how much could be allocated
    before swapping"; there is no portable equivalent, so elsewhere it is
    absent rather than approximated.
    """
    out: dict[str, Any] = {"total_bytes": None, "available_bytes": None}
    try:
        pages = os.sysconf("SC_PHYS_PAGES")
        page = os.sysconf("SC_PAGE_SIZE")
        out["total_bytes"] = pages * page
    except (ValueError, OSError):
        pass
    meminfo = Path("/proc/meminfo")
    if meminfo.exists():
        try:
            for line in meminfo.read_text().splitlines():
                if line.startswith("MemAvailable:"):
                    out["available_bytes"] = int(line.split()[1]) * 1024
                    break
        except (OSError, ValueError, IndexError):
            pass
    return out


def _user_state(con: sqlite3.Connection, user: users.User) -> str:
    """The same four states `api/admin.py:_state` computes, restated here
    rather than imported because that one is that module's private helper
    and this one feeds an aggregate — if the two ever disagree, the accounts
    table and the dashboard tile disagree visibly on the same page, which is
    the failure mode a shared import would also produce, one refactor later
    and harder to see."""
    if user.disabled:
        return "disabled"
    if users.has_password(con, user.id):
        return "active"
    if tokens.outstanding(con, user.id, tokens.Purpose.INVITE):
        return "invited"
    return "no password"


def _ago(days: int) -> str:
    return (datetime.now(UTC) - timedelta(days=days)).isoformat()


# ------------------------------------------------------------------- routes

def install(app: FastAPI) -> None:
    """Add the stats routes. Read-only, admin-only, all on-box."""

    @app.get("/api/admin/stats/system")
    def system_stats(caller: Admin) -> dict[str, Any]:
        """The process and the machine under it, as the box reports them."""
        del caller
        disk = shutil.disk_usage(config.DATA_DIR)
        try:
            load = list(os.getloadavg())
        except OSError:
            load = []
        return {
            "process": _rss(),
            "memory": _machine_memory(),
            "load": load,
            "cpus": os.cpu_count(),
            "disk": {"path": str(config.DATA_DIR), "total_bytes": disk.total,
                     "used_bytes": disk.used, "free_bytes": disk.free},
        }

    @app.get("/api/admin/stats/storage")
    def storage_stats(caller: Admin) -> dict[str, Any]:
        """What is on the volume, named the way the architecture names it.

        Sizes are bytes or null — null meaning "nothing there yet", which on
        a fresh instance is most of them and is information, not an error.
        """
        del caller
        cache = config.DATA_DIR / "cache"
        cache_bytes = _size_of(cache)
        return {
            "app_db_bytes": _size_of(config.APP_DB_PATH),
            "pool_bytes": _size_of(config.DB_PATH),
            "scryfall_bulk_bytes": _size_of(config.SCRYFALL_DIR),
            "cache_bytes": cache_bytes,
            "cache": _cache_breakdown(cache, cache_bytes),
            "decks": {
                "count": _count_dirs(config.DECKS_DIR),
                "bytes": _size_of(config.DECKS_DIR),
                "trashed": _count_dirs(config.DECKS_DIR / ".trash"),
            },
        }

    @app.get("/api/admin/stats/claude")
    def claude_stats(caller: Admin) -> dict[str, Any]:
        """Where the Claude tokens went — per mode, per model, and in dollars.

        The ledger's numbers (ADR 28's sibling in `claude/ledger.py`):
        aggregate counters only — no user, no deck, no question text. The
        caveat is part of the payload because it must ride with the numbers
        anywhere they are shown, the same rule the simulator's Tier 1
        caveats follow: cache *writes* are not captured, so these totals
        are a floor on the bill, never the bill.

        Two axes per window, because they answer different questions and
        neither substitutes for the other: **per mode** is "which surface is
        spending this", **per model** is "on which Claude" — the question
        per-account tiers made worth asking. Each is grouped in SQL rather
        than pivoted here, so a window's two lists sum to the same totals by
        construction.

        The **dollar figure is estimated from the per-model rollup only**, and
        that is not an optimisation. Pricing the per-mode one would mean
        pricing rows whose model column reads `(various)` — every rate would be
        a guess, and the guess would look like arithmetic.

        `estimated_usd` carries its own honesty with it: the date the rates
        were last checked by a person, and a count of conversations whose model
        this build cannot price. A caller rendering the figure without the
        second is showing a number that is wrong downward.
        """
        del caller

        def labelled(rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
            """Each row, plus how to name the model on a screen.

            Beside the id rather than instead of it. The label is what renders
            (commandment 10 — a model id is technology, and the tier picker two
            tabs over already says "Sonnet"); the id stays in the payload
            because it is what the pricing question is actually about, and a
            client that wanted to group or key on it should not have to parse
            prose to get it back.
            """
            return [{**row, "model_label": tiers.label_for(row["model"])}
                    for row in rows]

        def window(since: str | None) -> dict[str, Any]:
            by_mode = ledger_summary(since=since)
            by_model = ledger_summary(since=since, by="model")
            return {
                "by_mode": labelled(by_mode),
                "by_model": labelled(by_model),
                # Priced from the unlabelled rows: the estimate keys on the id,
                # and two ids that share a label (two Opus generations, say)
                # are two rates.
                "estimated_usd": prices.estimate(by_model).as_dict(),
            }

        return {
            "windows": {
                "week": window(_ago(7)),
                "month": window(_ago(30)),
                "all": window(None),
            },
            "caveat": "Token counts are a floor on the bill, not the bill: "
                      "cache writes bill at 1.25x input and are not captured.",
            "prices": {
                "checked": prices.CHECKED.isoformat(),
                "source": prices.SOURCE,
                "note": "Estimated from list rates read by a person on the "
                        "date above, not from an invoice. A conversation "
                        "whose model is not in the table is counted, never "
                        "priced at zero.",
            },
        }

    @app.get("/api/admin/stats/fly")
    async def fly_stats(caller: Admin) -> dict[str, Any]:
        """What the platform sees: Fly's managed Prometheus, if configured.

        The one view here that leaves the box, and the only one that can
        be switched off — `FLY_METRICS_TOKEN` is the maintainer's to mint,
        and unset means `configured: false` and a widget that hides rather
        than a dashboard that looks broken.

        `run_in_threadpool` because `flymetrics.fetch` is blocking stdlib
        `urllib` and this handler is `async`: called directly it would
        stall the event loop for every other request for as long as Fly
        takes to answer.
        """
        del caller
        from starlette.concurrency import run_in_threadpool

        from mtglab.api import flymetrics
        return await run_in_threadpool(flymetrics.fetch)

    @app.get("/api/admin/stats/traffic")
    def traffic_stats(caller: Admin) -> dict[str, Any]:
        """The visitor ledger: requests per day by status class, top routes.

        Route templates and status classes only — the recording side
        (`api/traffic.py`, schema v9) never held anything finer, so this
        view cannot leak what was never kept: no addresses, no agents, no
        names, no concrete paths. `note` rides in the payload the way the
        Claude view's caveat does, because the sentence belongs wherever
        the numbers go.
        """
        del caller
        from mtglab.api import traffic
        return {
            **traffic.summary(days=30),
            "note": "Route templates and status classes only — the ledger "
                    "never records an address, an agent, a name, or a "
                    "concrete path.",
        }

    @app.get("/api/admin/stats/activity")
    def activity_stats(caller: Admin) -> dict[str, Any]:
        """Who has been here and what the instance has been doing.

        Counts throughout — the visitor ledger (route-template request
        counts, schema v9) is a later PR on its own watched branch, so
        today this is what `app.db` already knows: accounts by state,
        sessions by recency, deck edits by day, memoised simulations, and
        the job registry's census.
        """
        del caller
        with db.connection() as con:
            states: dict[str, int] = {}
            for user in users.all_users(con):
                state = _user_state(con, user)
                states[state] = states.get(state, 0) + 1

            def one(query: str, args: tuple[Any, ...] = ()) -> int:
                row = con.execute(query, args).fetchone()
                return int(row[0]) if row and row[0] is not None else 0

            sessions_total = one("SELECT count(*) FROM sessions")
            seen_day = one(
                "SELECT count(*) FROM sessions WHERE last_seen_at >= ?",
                (_ago(1),))
            seen_week = one(
                "SELECT count(*) FROM sessions WHERE last_seen_at >= ?",
                (_ago(7),))
            edits = [
                {"day": row[0], "edits": row[1]}
                for row in con.execute(
                    "SELECT substr(created_at, 1, 10) AS day, count(*)"
                    " FROM deck_log WHERE created_at >= ?"
                    " GROUP BY day ORDER BY day",
                    (_ago(30),)).fetchall()
            ]
            sim_rows = one("SELECT count(*) FROM sim_cache")

        return {
            "accounts": states,
            "sessions": {"total": sessions_total, "seen_day": seen_day,
                         "seen_week": seen_week},
            "deck_edits_by_day": edits,
            "sim_cache_rows": sim_rows,
            "jobs": jobs.census(),
        }


def ledger_summary(*, since: str | None = None,
                   by: str = "mode") -> list[dict[str, Any]]:
    """`claude.ledger.summary`, imported lazily the way the runs modules
    import their planners: the ledger rides the base install, but keeping
    the import at call time keeps this module importable in any test that
    stubs it.

    `claude.prices` is imported at module level instead, and the difference is
    deliberate: this one touches the database and is the thing tests stub,
    while that one is a table and some arithmetic with nothing to fail.
    """
    from mtglab.claude import ledger
    return ledger.summary(since=since, by=by)

File: go/internal/api/adminstats.go
// The admin dashboard's numbers — `api/adminstats.py`, the last route family
// to cross (Phase 8). Everything lives under the admin prefix the middleware
// refuses before routing (ADR 17), and every handler also asks
// `requireAdmin`, the same two walls Python keeps.
//
// All the views are facts about this box, read from the box. Two couplings
// dissolved to let them cross: the job registry census is the door's own now
// (no Python route creates a job any more, so uvicorn's registry is always
// empty and this one is the honest total), and `_rss` reports the process on
// the port — which during the last of the coexistence is the door rather
// than the pair, and at retirement is simply the process.

package api

import (
	"database/sql"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"runtime"
	"time"

	"github.com/aasquier/sylvan-library/go/internal/auth"
	"github.com/aasquier/sylvan-library/go/internal/claude/ledger"
	"github.com/aasquier/sylvan-library/go/internal/floats"
	"github.com/aasquier/sylvan-library/go/internal/prices"
	"github.com/aasquier/sylvan-library/go/internal/tiers"
	"github.com/aasquier/sylvan-library/go/internal/wire"

	// The stats driver for the read-only schema peek; registered by the auth
	// package's import already, named here for clarity.
	_ "modernc.org/sqlite"
)

// statsSystem is `GET /api/admin/stats/system`: the process and the machine
// under it, as the box reports them.
func (a *API) statsSystem(w http.ResponseWriter, r *http.Request) {
	if a.requireAdmin(w, r) {
		return
	}
	rssBytes, rssKind := processRSS()
	memTotal, memAvailable := machineMemory()
	loads := loadAverages()
	loadOut := make([]any, 0, len(loads))
	for _, l := range loads {
		loadOut = append(loadOut, floats.Float(l))
	}
	total, used, free := diskUsage(a.dataDir)
	wire.JSON(w, http.StatusOK, wire.OrderedMap{
		{Key: "schema", Value: wire.OrderedMap{
			{Key: "applied", Value: a.schemaApplied()},
			{Key: "expected", Value: auth.SchemaVersion},
		}},
		{Key: "process", Value: wire.OrderedMap{
			{Key: "bytes", Value: rssBytes},
			{Key: "kind", Value: rssKind},
		}},
		{Key: "memory", Value: wire.OrderedMap{
			{Key: "total_bytes", Value: intOrNil(memTotal)},
			{Key: "available_bytes", Value: intOrNil(memAvailable)},
		}},
		{Key: "load", Value: loadOut},
		{Key: "cpus", Value: runtime.NumCPU()},
		{Key: "disk", Value: wire.OrderedMap{
			{Key: "path", Value: a.dataDir},
			{Key: "total_bytes", Value: total},
			{Key: "used_bytes", Value: used},
			{Key: "free_bytes", Value: free},
		}},
	})
}

// schemaApplied reads `user_version` off `app.db` through a bare read-only
// connection — never through the ladder, because a stats panel that could
// change the schema by being looked at is not a stats panel. `mode=ro`
// refuses to create the file and the exists() check keeps a bare laptop
// from logging a warning per refresh; the redundancy is deliberate, exactly
// as `adminstats._schema` argues it.
func (a *API) schemaApplied() any {
	if a.dbPath == "" {
		return nil
	}
	if _, err := os.Stat(a.dbPath); err != nil {
		return nil
	}
	db, err := sql.Open("sqlite", "file:"+url.PathEscape(a.dbPath)+"?mode=ro")
	if err != nil {
		a.log.Warn("could not read app.db schema version", "error", err)
		return nil
	}
	defer func() { _ = db.Close() }()
	var version int
	if err := db.QueryRow("PRAGMA user_version").Scan(&version); err != nil {
		a.log.Warn("could not read app.db schema version", "error", err)
		return nil
	}
	return version
}

// statsStorage is `GET /api/admin/stats/storage`: what is on the volume,
// named the way the architecture names it. Sizes are bytes or null — null
// meaning "nothing there yet", which on a fresh instance is most of them
// and is information, not an error.
func (a *API) statsStorage(w http.ResponseWriter, r *http.Request) {
	if a.requireAdmin(w, r) {
		return
	}
	cache := filepath.Join(a.dataDir, "cache")
	cacheBytes := sizeOf(cache)
	wire.JSON(w, http.StatusOK, wire.OrderedMap{
		{Key: "app_db_bytes", Value: intOrNil(sizeOf(a.dbPath))},
		{Key: "pool_bytes", Value: intOrNil(sizeOf(a.poolPath))},
		{Key: "scryfall_bulk_bytes", Value: intOrNil(sizeOf(a.scryfallDir))},
		{Key: "cache_bytes", Value: intOrNil(cacheBytes)},
		{Key: "cache", Value: cacheBreakdown(cache, cacheBytes)},
		{Key: "decks", Value: wire.OrderedMap{
			{Key: "count", Value: countDirs(a.decksDir)},
			{Key: "bytes", Value: intOrNil(sizeOf(a.decksDir))},
			{Key: "trashed", Value: countDirs(filepath.Join(a.decksDir, ".trash"))},
		}},
	})
}

// cacheBreakdown names the three shelves and whatever is left over as
// `other_bytes` — the part that matters, because this panel once shipped
// naming two of three tenants while the reading engine was 38% of the
// cache. A remainder cannot hide a fourth shelf the way a fixed list can.
func cacheBreakdown(cache string, total *int64) wire.OrderedMap {
	symbols := sizeOf(filepath.Join(cache, "symbols"))
	cardmotion := sizeOf(filepath.Join(cache, "cardmotion"))
	ocr := sizeOf(filepath.Join(cache, "ocr"))
	var other any
	if total != nil {
		named := int64(0)
		for _, v := range []*int64{symbols, cardmotion, ocr} {
			if v != nil {
				named += *v
			}
		}
		rest := *total - named
		if rest < 0 {
			rest = 0
		}
		other = rest
	}
	return wire.OrderedMap{
		{Key: "symbols_bytes", Value: intOrNil(symbols)},
		{Key: "cardmotion_bytes", Value: intOrNil(cardmotion)},
		{Key: "ocr_bytes", Value: intOrNil(ocr)},
		{Key: "other_bytes", Value: other},
	}
}

// sizeOf is `_size_of`: bytes on disk, or nil for "nothing there".
func sizeOf(path string) *int64 {
	if path == "" {
		return nil
	}
	info, err := os.Stat(path)
	if err != nil {
		return nil
	}
	if info.Mode().IsRegular() {
		size := info.Size()
		return &size
	}
	if !info.IsDir() {
		return nil
	}
	var total int64
	err = filepath.WalkDir(path, func(_ string, d os.DirEntry, err error) error {
		if err != nil || d.IsDir() {
			return nil //nolint:nilerr // a vanished entry is skipped, as rglob skips it
		}
		if fi, err := d.Info(); err == nil && fi.Mode().IsRegular() {
			total += fi.Size()
		}
		return nil
	})
	if err != nil {
		return nil
	}
	return &total
}

func countDirs(path string) int {
	entries, err := os.ReadDir(path)
	if err != nil {
		return 0
	}
	count := 0
	for _, e := range entries {
		if e.IsDir() {
			count++
		}
	}
	return count
}

func intOrNil(v *int64) any {
	if v == nil {
		return nil
	}
	return *v
}

// statsClaude is `GET /api/admin/stats/claude`: where the Claude tokens
// went — per mode, per model, and in dollars. Two axes per window because
// they answer different questions; the dollar figure is estimated from the
// per-model rollup only, since pricing per-mode rows would price
// `(various)` and the guess would look like arithmetic. The caveat rides in
// the payload because it must ride with the numbers anywhere they go.
func (a *API) statsClaude(w http.ResponseWriter, r *http.Request) {
	if a.requireAdmin(w, r) {
		return
	}
	// An instance with no app.db answers the shapes Python answers over the
	// empty database `db.connection()` would mint there: empty roll-ups,
	// never a 404 — the stats are about the instance, not about an account.
	db, present := a.accountsDB()
	rec := a.claudeLedger
	if rec == nil && present {
		rec = ledger.RecorderFrom(db, a.log)
	}
	window := func(since string) (wire.OrderedMap, error) {
		byMode, byModel := []ledger.Summary{}, []ledger.Summary{}
		if rec != nil {
			var err error
			if byMode, err = rec.Summarise(r.Context(), "mode", since); err != nil {
				return nil, err
			}
			if byModel, err = rec.Summarise(r.Context(), "model", since); err != nil {
				return nil, err
			}
		}
		priceRows := make([]prices.Row, 0, len(byModel))
		for _, row := range byModel {
			priceRows = append(priceRows, prices.Row{Model: row.Model,
				Conversations: int64(row.Conversations),
				InputTokens:   int64(row.InputTokens),
				OutputTokens:  int64(row.OutputTokens),
				CacheRead:     int64(row.CacheReadTokens)})
		}
		return wire.OrderedMap{
			{Key: "by_mode", Value: labelled(byMode, "mode")},
			{Key: "by_model", Value: labelled(byModel, "model")},
			{Key: "estimated_usd", Value: prices.Over(priceRows, prices.Today()).AsDict()},
		}, nil
	}
	week, err := window(ago(7))
	if err == nil {
		var month, all wire.OrderedMap
		if month, err = window(ago(30)); err == nil {
			all, err = window("")
		}
		if err == nil {
			wire.JSON(w, http.StatusOK, wire.OrderedMap{
				{Key: "windows", Value: wire.OrderedMap{
					{Key: "week", Value: week},
					{Key: "month", Value: month},
					{Key: "all", Value: all},
				}},
				{Key: "caveat", Value: "Token counts are a floor on the bill, " +
					"not the bill: cache writes bill at 1.25x input and are " +
					"not captured."},
				{Key: "prices", Value: wire.OrderedMap{
					{Key: "checked", Value: prices.Checked},
					{Key: "source", Value: prices.Source},
					{Key: "note", Value: "Estimated from list rates read by a " +
						"person on the date above, not from an invoice. A " +
						"conversation whose model is not in the table is " +
						"counted, never priced at zero."},
				}},
			})
			return
		}
	}
	a.fail(w, "stats/claude", err)
}

// labelled is each row plus how to name the model on a screen — beside the
// id, never instead of it (commandment 10: the label renders, the id is
// what the pricing question is about). The key order follows the axis: the
// grouped column first, then `(various)` — the SELECT order a `dict(row)`
// preserves in Python.
func labelled(rows []ledger.Summary, by string) []wire.OrderedMap {
	out := make([]wire.OrderedMap, 0, len(rows))
	for _, row := range rows {
		first := wire.KV{Key: "mode", Value: row.Mode}
		second := wire.KV{Key: "model", Value: row.Model}
		if by == "model" {
			first, second = second, first
		}
		out = append(out, wire.OrderedMap{
			first, second,
			{Key: "conversations", Value: row.Conversations},
			{Key: "requests", Value: row.Requests},
			{Key: "input_tokens", Value: row.InputTokens},
			{Key: "output_tokens", Value: row.OutputTokens},
			{Key: "cache_read_tokens", Value: row.CacheReadTokens},
			{Key: "first_at", Value: row.FirstAt},
			{Key: "last_at", Value: row.LastAt},
			{Key: "model_label", Value: tiers.LabelFor(row.Model)},
		})
	}
	return out
}

// ago is `_ago`: an ISO-8601 UTC instant `days` back, for string-compared
// `created_at` windows.
func ago(days int) string {
	return time.Now().UTC().AddDate(0, 0, -days).Format("2006-01-02T15:04:05.000000+00:00")
}

// statsFly is `GET /api/admin/stats/fly`: what the platform sees, when the
// platform is asked — the one view that leaves the box, and the only one
// that can be switched off.
func (a *API) statsFly(w http.ResponseWriter, r *http.Request) {
	if a.requireAdmin(w, r) {
		return
	}
	wire.JSON(w, http.StatusOK, a.fly.Fetch())
}

// statsTraffic is `GET /api/admin/stats/traffic`: the visitor ledger. The
// note rides in the payload the way the Claude view's caveat does, because
// the sentence belongs wherever the numbers go.
func (a *API) statsTraffic(w http.ResponseWriter, r *http.Request) {
	if a.requireAdmin(w, r) {
		return
	}
	summary, err := a.traffic.Summary(r.Context(), 30)
	if err != nil {
		a.fail(w, "stats/traffic", err)
		return
	}
	wire.JSON(w, http.StatusOK, append(summary,
		wire.KV{Key: "note", Value: "Route templates and status classes " +
			"only — the ledger never records an address, an agent, a name, " +
			"or a concrete path."}))
}

// statsActivity is `GET /api/admin/stats/activity`: who has been here and
// what the instance has been doing — accounts by state, sessions by
// recency, deck edits by day, memoised simulations, and the job registry's
// census.
func (a *API) statsActivity(w http.ResponseWriter, r *http.Request) {
	if a.requireAdmin(w, r) {
		return
	}
	// Like `statsClaude`: an absent app.db is Python's freshly-minted empty
	// one — zeroes and empty lists, never a 404.
	db, present := a.accountsDB()
	if !present {
		census := map[string]int{}
		if a.jobs != nil {
			census = a.jobs.Census()
		}
		wire.JSON(w, http.StatusOK, wire.OrderedMap{
			{Key: "accounts", Value: wire.OrderedMap{}},
			{Key: "sessions", Value: wire.OrderedMap{
				{Key: "total", Value: 0}, {Key: "seen_day", Value: 0},
				{Key: "seen_week", Value: 0},
			}},
			{Key: "deck_edits_by_day", Value: []any{}},
			{Key: "sim_cache_rows", Value: 0},
			{Key: "jobs", Value: census},
		})
		return
	}
	ctx := r.Context()
	users, err := auth.AllUsers(ctx, db)
	if err != nil {
		a.fail(w, "stats/activity", err)
		return
	}
	// First-encounter order over the username-sorted list, exactly the
	// order Python's dict grows in.
	states := wire.OrderedMap{}
	for _, u := range users {
		state, err := accountState(ctx, db, u)
		if err != nil {
			a.fail(w, "stats/activity", err)
			return
		}
		found := false
		for i := range states {
			if states[i].Key == state {
				states[i].Value = states[i].Value.(int64) + 1
				found = true
			}
		}
		if !found {
			states = append(states, wire.KV{Key: state, Value: int64(1)})
		}
	}

	one := func(query string, args ...any) (int64, error) {
		var n sql.NullInt64
		if err := db.QueryRowContext(ctx, query, args...).Scan(&n); err != nil {
			return 0, err
		}
		return n.Int64, nil
	}
	sessionsTotal, err := one("SELECT count(*) FROM sessions")
	if err == nil {
		var seenDay, seenWeek, simRows int64
		if seenDay, err = one("SELECT count(*) FROM sessions WHERE last_seen_at >= ?", ago(1)); err == nil {
			if seenWeek, err = one("SELECT count(*) FROM sessions WHERE last_seen_at >= ?", ago(7)); err == nil {
				simRows, err = one("SELECT count(*) FROM sim_cache")
			}
		}
		if err == nil {
			edits := []any{}
			rows, queryErr := db.QueryContext(ctx,
				"SELECT substr(created_at, 1, 10) AS day, count(*)"+
					" FROM deck_log WHERE created_at >= ?"+
					" GROUP BY day ORDER BY day", ago(30))
			if queryErr != nil {
				a.fail(w, "stats/activity", queryErr)
				return
			}
			defer func() { _ = rows.Close() }()
			for rows.Next() {
				var day string
				var count int64
				if err := rows.Scan(&day, &count); err != nil {
					a.fail(w, "stats/activity", err)
					return
				}
				edits = append(edits, wire.OrderedMap{
					{Key: "day", Value: day},
					{Key: "edits", Value: count},
				})
			}
			if err := rows.Err(); err != nil {
				a.fail(w, "stats/activity", err)
				return
			}
			census := map[string]int{}
			if a.jobs != nil {
				census = a.jobs.Census()
			}
			wire.JSON(w, http.StatusOK, wire.OrderedMap{
				{Key: "accounts", Value: states},
				{Key: "sessions", Value: wire.OrderedMap{
					{Key: "total", Value: sessionsTotal},
					{Key: "seen_day", Value: seenDay},
					{Key: "seen_week", Value: seenWeek},
				}},
				{Key: "deck_edits_by_day", Value: edits},
				{Key: "sim_cache_rows", Value: simRows},
				// A Python dict's order here is whatever order statuses were
				// first seen in an arbitrary registry walk — not a shape any
				// golden can hold — so the sorted rendering a Go map gets is
				// one of the orders Python itself produces.
				{Key: "jobs", Value: census},
			})
			return
		}
	}
	a.fail(w, "stats/activity", err)
}

// The four account states come from `accountState` in admin.go — Python
// restates `_user_state` beside `_state` so a disagreement is visible on one
// page, but the two are the same predicate and this port's admin surface
// already carries it; a second copy here would be the drift, not the guard.
